File: tests/docker-compat/python-client/tests/streams_advanced.py
"""Streams Advanced tests - consumer groups, pending, claiming, trimming."""

import time
from .framework import (
    register, run_test, key, assert_equal, assert_true,
    assert_equal_int,
)


def test_streams_advanced(r, prefix):
    results = []
    cat = "streams_advanced"

    # XGROUP_CREATE: Create consumer group, verify with XINFO GROUPS
    def t_xgroup_create():
        k = key(prefix, "sa:xgroup")
        try:
            r.xadd(k, {"f": "v"}, id="1-1")
            r.xgroup_create(k, "mygroup", "0")
            groups = r.xinfo_groups(k)
            assert_equal_int(1, len(groups))
            assert_equal("mygroup", groups[0]["name"])
        finally:
            r.delete(k)
    results.append(run_test("XGROUP_CREATE", cat, t_xgroup_create))

    # XREADGROUP_basic: Create group, XREADGROUP, verify message delivered
    def t_xreadgroup_basic():
        k = key(prefix, "sa:xreadgroup")
        try:
            r.xadd(k, {"msg": "hello"}, id="1-1")
            r.xgroup_create(k, "grp1", "0")
            res = r.xreadgroup("grp1", "consumer1", {k: ">"}, count=10)
            assert_equal_int(1, len(res))
            # res = [[stream_name, [(id, {field: value}), ...]]]
            messages = res[0][1]
            assert_equal_int(1, len(messages))
            assert_equal("hello", messages[0][1]["msg"])
        finally:
            r.delete(k)
    results.append(run_test("XREADGROUP_basic", cat, t_xreadgroup_basic))

    # XREADGROUP_pending: XREADGROUP without ACK, verify pending
    def t_xreadgroup_pending():
        k = key(prefix, "sa:xreadpend")
        try:
            r.xadd(k, {"f": "v"}, id="1-1")
            r.xadd(k, {"f": "v2"}, id="2-1")
            r.xgroup_create(k, "grp1", "0")
            r.xreadgroup("grp1", "c1", {k: ">"}, count=10)
            pending = r.xpending(k, "grp1")
            assert_equal_int(2, pending["pending"])
        finally:
            r.delete(k)
    results.append(run_test("XREADGROUP_pending", cat, t_xreadgroup_pending))

    # XACK_basic: Read message, XACK, verify removed from pending
    def t_xack_basic():
        k = key(prefix, "sa:xack")
        try:
            entry_id = r.xadd(k, {"f": "v"})
            r.xgroup_create(k, "grp1", "0")
            r.xreadgroup("grp1", "c1", {k: ">"}, count=10)

            # Verify pending before ACK
            pending = r.xpending(k, "grp1")
            assert_equal_int(1, pending["pending"])

            # ACK
            n = r.xack(k, "grp1", entry_id)
            assert_equal_int(1, n)

            # Verify pending after ACK
            pending = r.xpending(k, "grp1")
            assert_equal_int(0, pending["pending"])
        finally:
            r.delete(k)
    results.append(run_test("XACK_basic", cat, t_xack_basic))

    # XCLAIM_basic: Read on consumer1, XCLAIM to consumer2
    def t_xclaim_basic():
        k = key(prefix, "sa:xclaim")
        try:
            entry_id = r.xadd(k, {"f": "v"})
            r.xgroup_create(k, "grp1", "0")
            r.xreadgroup("grp1", "consumer1", {k: ">"}, count=10)

            # Claim to consumer2 with 0 min-idle-time
            msgs = r.xclaim(k, "grp1", "consumer2", 0, [entry_id])
            assert_equal_int(1, len(msgs))

            # Verify consumer2 has it in pending
            detail = r.xpending_range(k, "grp1", "-", "+", 10)
            assert_equal_int(1, len(detail))
            assert_equal("consumer2", detail[0]["consumer"])
        finally:
            r.delete(k)
    results.append(run_test("XCLAIM_basic", cat, t_xclaim_basic))

    # XAUTOCLAIM_basic: Read message, wait, XAUTOCLAIM with min-idle
    def t_xautoclaim_basic():
        k = key(prefix, "sa:xautoclaim")
        try:
            r.xadd(k, {"f": "v"}, id="1-1")
            r.xgroup_create(k, "grp1", "0")
            r.xreadgroup("grp1", "consumer1", {k: ">"}, count=10)

            # Small wait to ensure idle time > 0
            time.sleep(0.05)

            # XAUTOCLAIM to consumer2 with very low min-idle
            result = r.xautoclaim(k, "grp1", "consumer2", 1, "0-0", count=10)
            # result = (next_start_id, [(id, {fields}), ...])
            # or (next_start_id, [(id, {fields}), ...], [deleted_ids])
            claimed = result[1]
            assert_true(len(claimed) >= 1,
                        f"expected at least 1 autoclaimed message, got {len(claimed)}")
        finally:
            r.delete(k)
    results.append(run_test("XAUTOCLAIM_basic", cat, t_xautoclaim_basic))

    # XPENDING_summary: Check pending summary
    def t_xpending_summary():
        k = key(prefix, "sa:xpendsumm")
        try:
            r.xadd(k, {"f": "v1"}, id="1-1")
            r.xadd(k, {"f": "v2"}, id="2-1")
            r.xadd(k, {"f": "v3"}, id="3-1")
            r.xgroup_create(k, "grp1", "0")

            r.xreadgroup("grp1", "c1", {k: ">"}, count=2)
            r.xreadgroup("grp1", "c2", {k: ">"}, count=1)

            pending = r.xpending(k, "grp1")
            assert_equal_int(3, pending["pending"])
            assert_equal("1-1", pending["min"])
            assert_equal("3-1", pending["max"])
            # Should have 2 consumers
            assert_equal_int(2, len(pending["consumers"]))
        finally:
            r.delete(k)
    results.append(run_test("XPENDING_summary", cat, t_xpending_summary))

    # XPENDING_detail: Check per-consumer pending detail
    def t_xpending_detail():
        k = key(prefix, "sa:xpenddet")
        try:
            r.xadd(k, {"f": "v1"}, id="1-1")
            r.xadd(k, {"f": "v2"}, id="2-1")
            r.xgroup_create(k, "grp1", "0")
            r.xreadgroup("grp1", "c1", {k: ">"}, count=10)

            detail = r.xpending_range(k, "grp1", "-", "+", 10)
            assert_equal_int(2, len(detail))
            assert_equal("c1", detail[0]["consumer"])
            assert_equal("1-1", detail[0]["message_id"])
        finally:
            r.delete(k)
    results.append(run_test("XPENDING_detail", cat, t_xpending_detail))

    # XINFO_STREAM: Verify full stream info including length, groups, first/last entry
    def t_xinfo_stream():
        k = key(prefix, "sa:xinfost")
        try:
            r.xadd(k, {"f": "v1"}, id="1-1")
            r.xadd(k, {"f": "v2"}, id="2-1")
            r.xadd(k, {"f": "v3"}, id="3-1")
            r.xgroup_create(k, "grp1", "0")

            info = r.xinfo_stream(k)
            assert_equal_int(3, info["length"])
            assert_equal_int(1, info["groups"])
            assert_equal("1-1", info["first-entry"][0])
            assert_equal("3-1", info["last-entry"][0])
        finally:
            r.delete(k)
    results.append(run_test("XINFO_STREAM", cat, t_xinfo_stream))

    # XINFO_CONSUMERS: Verify consumer info
    def t_xinfo_consumers():
        k = key(prefix, "sa:xinfocons")
        try:
            r.xadd(k, {"f": "v"}, id="1-1")
            r.xadd(k, {"f": "v2"}, id="2-1")
            r.xgroup_create(k, "grp1", "0")
            r.xreadgroup("grp1", "consumer1", {k: ">"}, count=10)

            consumers = r.xinfo_consumers(k, "grp1")
            assert_equal_int(1, len(consumers))
            assert_equal("consumer1", consumers[0]["name"])
            assert_equal_int(2, consumers[0]["pending"])
        finally:
            r.delete(k)
    results.append(run_test("XINFO_CONSUMERS", cat, t_xinfo_consumers))

    # XTRIM_MAXLEN: Add entries, XTRIM MAXLEN 5, verify only 5 remain
    def t_xtrim_maxlen():
        k = key(prefix, "sa:xtrimmax")
        try:
            for i in range(10):
                r.xadd(k, {"i": str(i)})
            r.xtrim(k, maxlen=5, approximate=False)
            assert_equal_int(5, r.xlen(k))
        finally:
            r.delete(k)
    results.append(run_test("XTRIM_MAXLEN", cat, t_xtrim_maxlen))

    # XTRIM_MINID: Add entries, XTRIM MINID, verify entries before ID removed
    def t_xtrim_minid():
        k = key(prefix, "sa:xtrimmin")
        try:
            r.xadd(k, {"f": "v1"}, id="1-1")
            r.xadd(k, {"f": "v2"}, id="2-1")
            r.xadd(k, {"f": "v3"}, id="3-1")
            r.xadd(k, {"f": "v4"}, id="4-1")
            r.xadd(k, {"f": "v5"}, id="5-1")

            r.xtrim(k, minid="3-1")
            assert_equal_int(3, r.xlen(k))
            # Verify first entry is now 3-1
            msgs = r.xrange(k, "-", "+")
            assert_equal("3-1", msgs[0][0])
        finally:
            r.delete(k)
    results.append(run_test("XTRIM_MINID", cat, t_xtrim_minid))

    return results


register("streams_advanced", test_streams_advanced)

File: tests/docker-compat/js-client/tests/streams_advanced.js
const { register, runTest, key, assertEqual, assertTrue, assertEqualInt, sleep } = require('./framework');

register('streams_advanced', async (redis, prefix) => {
  const results = [];
  const cat = 'streams_advanced';

  // Helper to parse flat XINFO array into object
  function parseInfoArray(arr) {
    const obj = {};
    for (let i = 0; i < arr.length; i += 2) {
      obj[arr[i]] = arr[i + 1];
    }
    return obj;
  }

  // XGROUP_CREATE: Create consumer group, verify with XINFO GROUPS
  results.push(await runTest('XGROUP_CREATE', cat, async () => {
    const k = key(prefix, 'sa:xgroup');
    await redis.xadd(k, '1-1', 'f', 'v');
    await redis.xgroup('CREATE', k, 'mygroup', '0');
    const groups = await redis.call('XINFO', 'GROUPS', k);
    assertEqualInt(1, groups.length);
    const g = parseInfoArray(groups[0]);
    assertEqual('mygroup', g.name);
    await redis.del(k);
  }));

  // XREADGROUP_basic: Create group, XREADGROUP, verify message delivered
  results.push(await runTest('XREADGROUP_basic', cat, async () => {
    const k = key(prefix, 'sa:xreadgroup');
    await redis.xadd(k, '1-1', 'msg', 'hello');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    const res = await redis.xreadgroup('GROUP', 'grp1', 'consumer1', 'COUNT', 10, 'STREAMS', k, '>');
    assertEqualInt(1, res.length);
    assertEqualInt(1, res[0][1].length);
    // res[0][1][0] = [id, [field, value, ...]]
    assertEqual('hello', res[0][1][0][1][1]);
    await redis.del(k);
  }));

  // XREADGROUP_pending: XREADGROUP without ACK, verify pending
  results.push(await runTest('XREADGROUP_pending', cat, async () => {
    const k = key(prefix, 'sa:xreadpend');
    await redis.xadd(k, '1-1', 'f', 'v');
    await redis.xadd(k, '2-1', 'f', 'v2');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    await redis.xreadgroup('GROUP', 'grp1', 'c1', 'COUNT', 10, 'STREAMS', k, '>');
    const pending = await redis.xpending(k, 'grp1');
    // pending = [count, minId, maxId, [[consumer, count], ...]]
    assertEqualInt(2, pending[0]);
    await redis.del(k);
  }));

  // XACK_basic: Read message, XACK, verify removed from pending
  results.push(await runTest('XACK_basic', cat, async () => {
    const k = key(prefix, 'sa:xack');
    const id = await redis.xadd(k, '*', 'f', 'v');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    await redis.xreadgroup('GROUP', 'grp1', 'c1', 'COUNT', 10, 'STREAMS', k, '>');

    // Verify pending before ACK
    let pending = await redis.xpending(k, 'grp1');
    assertEqualInt(1, pending[0]);

    // ACK
    const n = await redis.xack(k, 'grp1', id);
    assertEqualInt(1, n);

    // Verify pending after ACK
    pending = await redis.xpending(k, 'grp1');
    assertEqualInt(0, pending[0]);
    await redis.del(k);
  }));

  // XCLAIM_basic: Read on consumer1, XCLAIM to consumer2
  results.push(await runTest('XCLAIM_basic', cat, async () => {
    const k = key(prefix, 'sa:xclaim');
    const id = await redis.xadd(k, '*', 'f', 'v');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    await redis.xreadgroup('GROUP', 'grp1', 'consumer1', 'COUNT', 10, 'STREAMS', k, '>');

    // Claim to consumer2 with 0 min-idle-time
    const msgs = await redis.xclaim(k, 'grp1', 'consumer2', 0, id);
    assertEqualInt(1, msgs.length);

    // Verify consumer2 has it in pending
    const detail = await redis.call('XPENDING', k, 'grp1', '-', '+', '10');
    assertEqualInt(1, detail.length);
    assertEqual('consumer2', detail[0][1]);
    await redis.del(k);
  }));

  // XAUTOCLAIM_basic: Read message, wait, XAUTOCLAIM with min-idle
  results.push(await runTest('XAUTOCLAIM_basic', cat, async () => {
    const k = key(prefix, 'sa:xautoclaim');
    await redis.xadd(k, '1-1', 'f', 'v');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    await redis.xreadgroup('GROUP', 'grp1', 'consumer1', 'COUNT', 10, 'STREAMS', k, '>');

    // Small wait to ensure idle time > 0
    await sleep(50);

    // XAUTOCLAIM to consumer2 with very low min-idle
    const result = await redis.call('XAUTOCLAIM', k, 'grp1', 'consumer2', '1', '0-0', 'COUNT', '10');
    // result = [nextStartId, [[id, [field, value, ...]], ...], deletedIds]
    assertTrue(result[1].length >= 1, `expected at least 1 autoclaimed message, got ${result[1].length}`);
    await redis.del(k);
  }));

  // XPENDING_summary: Check pending summary
  results.push(await runTest('XPENDING_summary', cat, async () => {
    const k = key(prefix, 'sa:xpendsumm');
    await redis.xadd(k, '1-1', 'f', 'v1');
    await redis.xadd(k, '2-1', 'f', 'v2');
    await redis.xadd(k, '3-1', 'f', 'v3');
    await redis.xgroup('CREATE', k, 'grp1', '0');

    await redis.xreadgroup('GROUP', 'grp1', 'c1', 'COUNT', 2, 'STREAMS', k, '>');
    await redis.xreadgroup('GROUP', 'grp1', 'c2', 'COUNT', 1, 'STREAMS', k, '>');

    const pending = await redis.xpending(k, 'grp1');
    // pending = [count, minId, maxId, [[consumer, count], ...]]
    assertEqualInt(3, pending[0]);
    assertEqual('1-1', pending[1]);
    assertEqual('3-1', pending[2]);
    // Should have 2 consumers
    assertEqualInt(2, pending[3].length);
    await redis.del(k);
  }));

  // XPENDING_detail: Check per-consumer pending detail
  results.push(await runTest('XPENDING_detail', cat, async () => {
    const k = key(prefix, 'sa:xpenddet');
    await redis.xadd(k, '1-1', 'f', 'v1');
    await redis.xadd(k, '2-1', 'f', 'v2');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    await redis.xreadgroup('GROUP', 'grp1', 'c1', 'COUNT', 10, 'STREAMS', k, '>');

    const detail = await redis.call('XPENDING', k, 'grp1', '-', '+', '10');
    assertEqualInt(2, detail.length);
    // Each entry: [id, consumer, idleTime, deliveryCount]
    assertEqual('c1', detail[0][1]);
    assertEqual('1-1', detail[0][0]);
    await redis.del(k);
  }));

  // XINFO_STREAM: Verify full stream info including length, groups, first/last entry
  results.push(await runTest('XINFO_STREAM', cat, async () => {
    const k = key(prefix, 'sa:xinfost');
    await redis.xadd(k, '1-1', 'f', 'v1');
    await redis.xadd(k, '2-1', 'f', 'v2');
    await redis.xadd(k, '3-1', 'f', 'v3');
    await redis.xgroup('CREATE', k, 'grp1', '0');

    const info = await redis.call('XINFO', 'STREAM', k);
    const obj = parseInfoArray(info);
    assertEqualInt(3, obj.length);
    assertEqualInt(1, obj.groups);
    // first-entry and last-entry are arrays: [id, [field, value, ...]]
    assertEqual('1-1', obj['first-entry'][0]);
    assertEqual('3-1', obj['last-entry'][0]);
    await redis.del(k);
  }));

  // XINFO_CONSUMERS: Verify consumer info
  results.push(await runTest('XINFO_CONSUMERS', cat, async () => {
    const k = key(prefix, 'sa:xinfocons');
    await redis.xadd(k, '1-1', 'f', 'v');
    await redis.xadd(k, '2-1', 'f', 'v2');
    await redis.xgroup('CREATE', k, 'grp1', '0');
    await redis.xreadgroup('GROUP', 'grp1', 'consumer1', 'COUNT', 10, 'STREAMS', k, '>');

    const consumers = await redis.call('XINFO', 'CONSUMERS', k, 'grp1');
    assertEqualInt(1, consumers.length);
    const c = parseInfoArray(consumers[0]);
    assertEqual('consumer1', c.name);
    assertEqualInt(2, c.pending);
    await redis.del(k);
  }));

  // XTRIM_MAXLEN: Add entries, XTRIM MAXLEN 5, verify only 5 remain
  results.push(await runTest('XTRIM_MAXLEN', cat, async () => {
    const k = key(prefix, 'sa:xtrimmax');
    for (let i = 0; i < 10; i++) {
      await redis.xadd(k, '*', 'i', String(i));
    }
    await redis.xtrim(k, 'MAXLEN', 5);
    const length = await redis.xlen(k);
    assertEqualInt(5, length);
    await redis.del(k);
  }));

  // XTRIM_MINID: Add entries, XTRIM MINID, verify entries before ID removed
  results.push(await runTest('XTRIM_MINID', cat, async () => {
    const k = key(prefix, 'sa:xtrimmin');
    await redis.xadd(k, '1-1', 'f', 'v1');
    await redis.xadd(k, '2-1', 'f', 'v2');
    await redis.xadd(k, '3-1', 'f', 'v3');
    await redis.xadd(k, '4-1', 'f', 'v4');
    await redis.xadd(k, '5-1', 'f', 'v5');

    await redis.xtrim(k, 'MINID', '3-1');
    const length = await redis.xlen(k);
    assertEqualInt(3, length);
    // Verify first entry is now 3-1
    const msgs = await redis.xrange(k, '-', '+');
    assertEqual('3-1', msgs[0][0]);
    await redis.del(k);
  }));

  return results;
});
